cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/HierarchyTreeNode/index.tsx

```diff
@@ -0,0 +1,338 @@
+      position: 'relative',
+      borderRadius: '3px',
+    },
+  };
+})(Paper);
+
+const AutocompleteContainer = withStyles(() => {
+  return {
+    root: {
+      padding: '2px 10px 2px 0px',
+      marginTop: '2px',
+      position: 'relative',
+      borderRadius: '3px',
+    },
+  };
+})(Paper);
+
+const RowButtonsWraper = withStyles(() => {
+  return {
+    root: {
+      display: 'grid',
+      gridTemplateColumns: '1fr 1fr',
+      justifyItems: 'end',
+    },
+  };
+})(Box);
+
+const ButtonIcon = withStyles(() => {
+  return {
+    root: {
+      padding: '6.5px',
+    },
+  };
+})(IconButton);
+
+const HiddenBtn = withStyles(() => {
+  return {
+    root: {
+      padding: '6.5px',
+      visibility: 'hidden',
+    },
+  };
+})(IconButton);
+
+interface IHierarchyTreeNodeStyleProps {
+  margin: number;
+}
+
+const useStyles = makeStyles<Theme, IHierarchyTreeNodeStyleProps>(() => {
+  return {
+    hierarchyIcon: {
+      width: '10px',
+      height: '34px',
+      position: 'absolute',
+      borderLeft: '2px solid rgba(0, 0, 0, 0.2)',
+      left: '-10px',
+    },
+    innerMostSiblingConnector: {
+      top: '17px',
+      left: '-8px',
+      width: '8px',
+      height: '2px',
+      content: '""',
+      position: 'absolute',
+      borderTop: '2px solid rgba(0, 0, 0, 0.2)',
+    },
+    hierarchyTreeMargin: {
+      marginLeft: (props) => props.margin,
+    },
+    multiSelectCustomMargin: {
+      marginLeft: (props) => (props.margin === 0 ? `10px` : `${props.margin}px`),
+    },
+    relative: {
+      position: 'relative',
+    },
+    checkBox: {
+      display: 'flex',
+      width: '100%',
+    },
+    paddingTextField: {
+      paddingLeft: '10px',
+    },
+  };
+});
+
+interface INodeProps {
+  id: number;
+  parentId?: number;
+  children?: INodeProps[];
+  name: string;
+  path?: string[];
+  type: string;
+}
+interface IHierarchyTreeNodeProps extends IHierarchyProps {
+  node: INodeProps;
+  setRecords: (value) => void | React.Dispatch<any>;
+  setTree: (value) => void | React.Dispatch<any>;
+  records: INodeProps[];
+  dropdownOptions: INodeProps[];
+  disabled: boolean;
+}
+
+const HierarchyTreeNode = ({
+  node,
+  setRecords,
+  setTree,
+  onChange,
+  records,
+  dropdownOptions,
+  disabled,
+}: IHierarchyTreeNodeProps) => {
+  const [multiSelectOptions, setMultiSelectOptions] = React.useState([]);
+  const [childVisible, setChildVisiblity] = React.useState(false);
+  const [showMultiSelect, setShowMultiSelect] = React.useState(false);
+  const hasChild = node.children.length !== 0 ? true : false;
+
+  let mrgL = 0;
+  if (hasChild && childVisible) {
+    mrgL += 10;
+  }
+
+  const classes = useStyles({ margin: mrgL });
+
+  const fillMultiSelectOptions = (e) => {
+    setMultiSelectOptions(e);
+  };
+
+  const handleMultiSelectClose = () => {
+    const newRecords = [...records];
+    multiSelectOptions.forEach((item) => {
+      let childrenOfSelected = [];
+
+      const uniqueId = uuidV4();
```

Comment:
As mentioned in the HierarchyWidget generate this uuid once when generating the dropdown options. That way it is done only once and not each time the user selects a field.